File: types/goals.ts
export type GoalPriority = 'high' | 'medium' | 'low';
export type GoalCategory = 'health' | 'fitness' | 'nutrition' | 'mental' | 'personal';

export interface TimeEstimate {
  min_duration: string;
  max_duration: string;
  recommended_frequency: string;
}

export interface WeeklyCompletionStatus {
  week_start: string;
  is_complete: boolean;
}

export interface ActionItem {
  title: string;
  description: string;
  priority: string;
  time_estimate: TimeEstimate;
  prerequisites: string[];
  success_criteria: string[];
  adaptation_notes: string[];
  weekly_completion?: WeeklyCompletionStatus[];
}

export interface ActionPlan {
  goal_id: string;
  goal_title: string;
  action_items: ActionItem[];
  total_estimated_time_per_week: string;
  health_adaptations: string[];
  suggested_schedule?: any;
}

export interface WeeklySchedule {
  start_date: string;
  end_date: string;
  daily_schedules: {
    [key: string]: {
      date: string;
      time_slots: Array<{
        start_time: string;
        end_time: string;
        duration: string;
        pillar?: string;
        action_item?: string;
        frequency?: string;
        priority?: string;
        health_notes?: string[];
      }>;
      total_duration: string;
      pillars_covered: string[];
    };
  };
  total_weekly_hours: number;
  pillar_distribution: { [key: string]: number };
  health_adaptations: string[];
  schedule_notes?: string[];
}

export interface GoalBase {
  title: string;
  description?: string;
  priority: GoalPriority;
  category: GoalCategory;
  target_value?: number;
  unit?: string;
  due_date?: string;
}

export interface GoalCreate extends GoalBase {
  user_email: string;
}

export interface GoalUpdate {
  title?: string;
  description?: string;
  priority?: GoalPriority;
  category?: GoalCategory;
  target_value?: number;
  unit?: string;
  current_value?: number;
  completed?: boolean;
  due_date?: string;
}

export interface Goal extends GoalBase {
  id: string;
  user_email: string;
  current_value?: number;
  completed: boolean;
  notes: string[];
  created_at: string;
  updated_at: string;
}

export interface GoalProgressUpdate {
  goal_id: string;
  current_value: number;
  note?: string;
}

export interface TimeEstimate {
  min_duration: string; // ISO duration
  max_duration: string; // ISO duration
  recommended_frequency: string;
}

export interface ActionItem {
  title: string;
  description: string;
  priority: string;
  time_estimate: TimeEstimate;
  prerequisites: string[];
  success_criteria: string[];
  adaptation_notes: string[];
}

export interface ActionPlan {
  goal_id: string;
  goal_title: string;
  action_items: ActionItem[];
  total_estimated_time_per_week: string; // ISO duration
  health_adaptations: string[];
  suggested_schedule?: any;
}

export interface GoalNote {
  goal_id: string;
  note: string;
}

export interface WeeklyReflection {
  user_email: string;
  week_start: string;
  week_end: string;
  rating: number;
  reflection?: string;
  next_week_goals: string[];
}

export interface WeeklyProgress {
  user_email: string;
  week_start: string;
  week_end: string;
  goals: Goal[];
  reflection?: string;
  rating?: number;
  created_at: string;
}

export interface HabitGoal {
  goal_id: string;
  habit_title: string;
  completed: boolean;
  completed_at?: string;
}

export interface GoalStats {
  total_goals: number;
  completed_goals: number;
  completion_rate: number;
  average_rating?: number;
  weekly_streak: number;
}

export interface GoalResponse {
  success: boolean;
  message: string;
  data?: any;
}

// Frontend specific types
export interface GoalFormData {
  title: string;
  description: string;
  priority: GoalPriority;
  category: GoalCategory;
  targetValue: string;
  unit: string;
  dueDate: Date;
}

export interface WeeklyProgressData {
  weekStart: Date;
  weekEnd: Date;
  goals: Goal[];
  reflection?: string;
  rating?: number;
  stats: {
    total_goals: number;
    completed_goals: number;
    completion_rate: number;
  };
}

File: evra-server/api/app/schemas/ai/goals.py
from pydantic import BaseModel, EmailStr, Field
from typing import List, Optional
from datetime import datetime, time
from enum import Enum
from app.schemas.backend.encrypt import EncryptedField

class GoalPriority(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"

class GoalCategory(str, Enum):
    HEALTH = "health"
    FITNESS = "fitness"
    NUTRITION = "nutrition"
    MENTAL = "mental"
    PERSONAL = "personal"

class GoalUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    priority: Optional[GoalPriority] = None
    category: Optional[GoalCategory] = None
    target_value: Optional[float] = Field(None, ge=0)
    unit: Optional[str] = Field(None, max_length=50)
    current_value: Optional[float] = Field(None, ge=0)
    completed: Optional[bool] = None
    due_date: Optional[datetime] = None

class GoalPriority(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"

# TODO: Covert user email to user_id
class GoalCreate(BaseModel):
    user_email: Optional[EmailStr] = None  # Set from token, not from request
    title: str = EncryptedField(...) 
    description: Optional[str] = EncryptedField(None)
    priority: GoalPriority
    category: GoalCategory

class Goal(GoalCreate):
    id: str
    created_at: datetime

class GoalProgressUpdate(BaseModel):
    goal_id: str
    current_value: float = Field(..., ge=0)
    note: Optional[str] = Field(None, max_length=500)

class GoalNote(BaseModel):
    goal_id: str
    note: str = Field(..., min_length=1, max_length=500)

# TODO: Change user_email to user_id
# TODO: Unable to encrypt integer fields
class WeeklyReflectionCreate(BaseModel):
    user_email: Optional[EmailStr] = None  # Set from token, not from request
    rating: int = Field(..., ge=1, le=5)
    reflection: Optional[str] = EncryptedField(None, max_length=2000)

class WeeklyReflection(WeeklyReflectionCreate):
    id: str
    created_at: datetime

class WeeklyProgress(BaseModel):
    user_email: EmailStr
    week_start: datetime
    week_end: datetime
    goals: List[Goal]
    reflection: Optional[str] = None
    rating: Optional[int] = Field(None, ge=1, le=5)
    created_at: datetime

class HabitGoal(BaseModel):
    goal_id: str
    habit_title: str = Field(..., min_length=1, max_length=200)
    completed: bool = False
    completed_at: Optional[datetime] = None

class GoalStats(BaseModel):
    total_goals: int
    completed_goals: int
    completion_rate: float
    average_rating: Optional[float] = None
    weekly_streak: int = 0
    total_weekly_streak_count: int = 0

class GoalResponse(BaseModel):
    success: bool
    message: str
    data: Optional[dict] = None 

# Action Plan Schemas

class ActionPriority(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"

# TODO: Change date, start_time, end_time to datetime format
# TODO: Notes field is not getting encrypted in Generate Plan flow
class DailySchedule(BaseModel):
    date: datetime
    start_time: Optional[datetime]
    end_time: Optional[datetime]
    notes: Optional[str] = EncryptedField(None)
    complete: bool = False

class WeeklyActionSchedule(BaseModel):
    monday: Optional[DailySchedule] = None
    tuesday: Optional[DailySchedule] = None
    wednesday: Optional[DailySchedule] = None
    thursday: Optional[DailySchedule] = None
    friday: Optional[DailySchedule] = None
    saturday: Optional[DailySchedule] = None
    sunday: Optional[DailySchedule] = None

# TODO: Remove priority from action items as it's already there in goal schema, convert user_email to user_id
class ActionItemCreate(BaseModel):
    goal_id: str
    title: str = EncryptedField(...)
    description: str = EncryptedField(...)
    priority: ActionPriority
    weekly_schedule: WeeklyActionSchedule

class ActionItem(ActionItemCreate):
    id: str
    user_email: Optional[EmailStr] = None  

class GoalWithActionItems(Goal):
    action_items: List[ActionItem] = []

class StreakScore(BaseModel):
    score: float
    week: int
